Sources/NIO/Codec.swift

```diff
@@ -96,64 +106,95 @@ private extension ChannelHandlerContext {
 
 extension ByteToMessageDecoder {
 
+    /// Decode in a loop until there is nothing more to decode.
+    private func decodeLoop(ctx: ChannelHandlerContext, decodeFunc: (ChannelHandlerContext, inout ByteBuffer) throws -> DecodingState) throws {
+        while var slice = self.cumulationBuffer?.slice(), slice.readableBytes > 0 {
```

Comment:
@weissi @Lukasa so I started to write some unit tests and notice we can not do this.. The problem here is that it is possible that `decodeLast` will be triggered from within `decodeFunc` (due a `close(...)` call for example). In this case `decodeLast` will by default call `decode(...)` which will then see the same bytes again even if `decode(...)` before increased the `readerIndex` as we had no chance to replicate these changes to the `cumulationBuffer` yet :(

Reply:
Yeah, there's no way to entirely prevent re-entrancy without removing decodeLast.

Reply:
yeah... so I wonder what we should do here in the case... any suggestions ?

Reply:
I mean, I'm inclined to sit on it and wait until 2.0, when decodeLast is removed. In the meantime we can recommend users override decodeLast or channelInactive.

Reply:
I mean, it doesn't matter what you do, the issue is still the same: `decodeLast` is not safe from avoid reentrancy. So you may as well keep using the slice.